### PR TITLE
Add generic inference for multiple pipelines

### PR DESCRIPTION
**Pull request by osanseviero**
This PR enables generic inference for different pipelines. I also created [templates](https://huggingface.co/templates).

Some examples I used to test the limits of the system:
* Audio to Audio: [Asteroid speech enhancement example](https://huggingface.co/osanseviero/ConvTasNet_Libri1Mix_enhsingle_16k)
* Automatic Speech Recognition: [pyctdecode + transformers ASR](https://huggingface.co/osanseviero/pyctcdecode_asr)
* Feature Extraction: [Fasttext](https://huggingface.co/osanseviero/fasttext_english). 
* Image Classification: [FastAI example dog vs cat classifier](https://huggingface.co/osanseviero/fastai_cat_vs_dog)
* Structured Data Classification: [sklearn wine (clone of julien)](https://huggingface.co/osanseviero/wine-quality)

TODO: Add tests to Docker and run the e2e test

FYI @nateraw @mishig25

### CODE REVIEW

**Review by lewtun — APPROVED** [review]
LGTM and very cool features 🚀 !

One question: what should an end-user do with the template repos? As far as I know we can't fork them on the Hub, so are they mainly there as a reference?

**Comment by julien-c** [comment]
@lewtun You can ’fork’ them with

```bash
git clone https://huggingface.co/templates/$TEMPLATE
cd $TEMPLATE
git remote set-url origin https://huggingface.co/lewtun/$TEMPLATE
git push [--force]
```

no?

**Comment by osanseviero** [comment]
Pretty much what Julien just said. You could also add `--depth 1` so it does not copy the whole repo history.

That's a good question and we should probably add a **how to clone** section in the card.

**Comment by lewtun** [comment]
yes, adding @julien-c's code snippet in the _how to clone_ button would be very useful!

File: api-inference-community/docker_images/generic/tests/test_api_audio_to_audio.py [hunk 1]
```diff
@@ -64,7 +70,7 @@ def test_simple(self):
         self.assertEqual(set(audio[0].keys()), {"blob", "content-type", "label"})
 
         data = base64.b64decode(audio[0]["blob"])
-        wavform = ffmpeg_read(data)
+        wavform = ffmpeg_read(data, 16000)
```
Comment (Narsil) [inline api-inference-community/docker_images/generic/tests/test_api_audio_to_audio.py:73]: Shouldn't we change `common` too on this one ?

File: api-inference-community/docker_images/generic/app/pipelines/audio_to_audio.py [hunk 2]
```diff
@@ -32,7 +28,4 @@ def __call__(self, inputs: np.array) -> Tuple[np.array, int, List[str]]:
                     This can be the name of the instruments for audio source separation
                     or some annotation for speech enhancement. The length must be `C'`.
         """
-        # IMPLEMENT_THIS
-        raise NotImplementedError(
-            "Please implement AudioToAudioPipeline __call__ function"
-        )
+        return super().__call__(inputs)
```
Comment (Narsil) [inline api-inference-community/docker_images/generic/app/pipelines/audio_to_audio.py:31]: I can't find the actual logic. Where is it ?

Comment (osanseviero) [reply api-inference-community/docker_images/generic/app/pipelines/audio_to_audio.py:31]: The main logic is in the `Pipeline` class
https://github.com/huggingface/huggingface_hub/blob/main/api-inference-community/docker_images/generic/app/pipelines/base.py#L10-L33

**Comment by osanseviero** [comment]
Hi @Narsil, I've done a high-level cleanup to use a single `Pipeline` instead of multiple repeating same code. I hope this looks better!

**Comment by Narsil** [comment]
This looks SSOOOOO much better ! and with `pip`c ache to boot, very nice :)

**Pull request merged**